[ModuleInterfaces] Rebuild if a dependency is missing

harlanhaskins opened the pull request:
A dependency being removed is not a hard error, because the recovery step is to recompile the interface. That might fail because of the change in dependencies, but that is a more actionable error.

Fixes rdar://51959698

Also, if there’s a `.swiftmodule` corresponding to the `.swiftinterface` we’re loading, but it’s out of date or otherwise unusable, still make it a dependency of the cached module.

The general assumption with cached modules is that, by virtue of them being in the cache, they have a more up-to-date view of the world than the adjacent compiled module. However, if a new `.swiftmodule` comes along that’s totally valid, we should consider it as having a newer view of the world than the cached one, so we should use it instead.

Fixes rdar://51959788

harlanhaskins committed:
[ModuleInterfaces] Rebuild if a dependency is missing

A dependency being removed is not a hard error, because the recovery step is to recompile the interface. That might fail because of the change in dependencies, but that is a more actionable error.

Fixes rdar://51959698

harlanhaskins committed:
[ModuleInterfaces] Add adjacent swiftmodule to dependencies

If there’s a `.swiftmodule` corresponding to the `.swiftinterface` we’re loading, but it’s out of date or otherwise unusable, still make it a dependency of the cached module.

The general assumption with cached modules is that, by virtue of them being in the cache, they have a more up-to-date view of the world than the adjacent compiled module. However, if a new `.swiftmodule` comes along that’s totally valid, we should consider it as having a newer view of the world than the cached one, so we should use it instead.

Fixes rdar://51959788

harlanhaskins committed:
[test] Test that missing dependencies cause a rebuild

Also make that dependency a .swiftmodule itself.

jrose-apple commented on lib/Frontend/ParseableInterfaceModuleLoader.cpp:
```diff
@@ -1294,6 +1337,14 @@ class ParseableInterfaceModuleLoaderImpl {
 
+    // If we found an out-of-date .swiftmodule, we still want to add it as
+    // a dependency of the .swiftinterface. That way if it's updated, but
+    // the .swiftinerface remains the same, we invalidate the cache and
```
Typo: "swiftinterface"

jrose-apple commented on test/ParseableInterface/ModuleCache/RebuildRemarks/missing-dependency.swift:
```diff
@@ -0,0 +1,26 @@
+// RUN: %target-swift-frontend -emit-module -o %t/Build/TestModule.swiftmodule %t/TestModule.swift -emit-module-interface-path %t/Build/TestModule.swiftinterface -swift-version 5 -module-name TestModule
+
+// 3. Make the compiled module unreadable so it gets added as a dependency but is not used
+// RUN: echo ' ' | cat - %t/Build/TestModule.swiftmodule > %t/Build/TestModule.swiftmodule.edited
```
`echo ' ' >> %t/Build/TestModule.swiftmodule`?

harlanhaskins replied:
I assumed that would still pass the sniff test, which tests the beginning of the file. I'll try by adding to the end of the file, though.

jrose-apple replied:
Oh, I get it. You could also just *replace* the file, I guess.

jrose-apple commented on test/ParseableInterface/ModuleCache/RebuildRemarks/missing-dependency.swift:
```diff
@@ -0,0 +1,26 @@
+// 4. Try to import the module, which will create a cached module that depends on both the .swiftmodule and .swiftinterface
+// RUN: %target-swift-frontend -typecheck %s -I %t/Build -module-cache-path %t/ModuleCache -Xllvm -debug-only=textual-module-interface
+
+// 5. Remove the .swiftmodule, which will result in a missing dependency note
```
I think the test should test the expected use case, which is that when you put the module back the way it was, the interface isn't used despite there being an entry in the cache.

harlanhaskins committed:
[ModuleInterfaces] Clear dependencies before checking a different module

Sometimes, if one module we tried failed, we would leave the dependencies from the previous check and fail a different module which would have succeeded.

This manifested as the prebuilt-module-cache-forwarding test testing the wrong thing, because it retained a mod-time based dependency in the prebuilt cache, which should have been hash based, and used it.

harlanhaskins commented:
@swift-ci please smoke test

harlanhaskins commented:
Gonna re-test just to make sure we're good to merge with the changes from the weekend.

@swift-ci please smoke test
